src/decode/libipt/mod.rs

```diff
@@ -0,0 +1,401 @@
+//! The libipt trace decoder.
+
+use crate::{c_errors::PerfPTCError, decode::TraceDecoder, errors::HWTracerError, Block, Trace};
+use libc::{c_char, c_int, c_void};
+use std::{convert::TryFrom, env, ffi::CString, os::fd::AsRawFd, ptr};
+use tempfile::NamedTempFile;
+
+extern "C" {
+    // decode.c
+    fn hwt_ipt_init_block_decoder(
+        buf: *const c_void,
+        len: u64,
+        vdso_fd: c_int,
+        vdso_filename: *const c_char,
+        decoder_status: *mut c_int,
+        err: *mut PerfPTCError,
+        current_exe: *const c_char,
+    ) -> *mut c_void;
+    fn hwt_ipt_next_block(
+        decoder: *mut c_void,
+        decoder_status: *mut c_int,
+        addr: *mut u64,
+        len: *mut u64,
+        err: *mut PerfPTCError,
+    ) -> bool;
+    fn hwt_ipt_free_block_decoder(decoder: *mut c_void);
+    // util.c
+    pub(crate) fn hwt_ipt_is_overflow_err(err: c_int) -> bool;
+    // libipt
+    pub(crate) fn pt_errstr(error_code: c_int) -> *const c_char;
+}
+
+pub(crate) struct LibIPTTraceDecoder {
+    //trace: &'t dyn Trace,
+}
+
+impl TraceDecoder for LibIPTTraceDecoder {
+    fn new() -> Self {
+        Self {}
+    }
+
+    fn iter_blocks<'t>(
```

Comment:
I claim that this can be `fn iter_blocks(&self, &dyn Trace) -> Box<dyn Iterator <Item = Result<Block, HWTracerError>>>`. I might be wrong!

Reply:
Just following compiler's orders guv'ner:

```
$ cargo check
    Checking hwtracer v0.1.0 (/home/vext01/research/hwtracer)
error: lifetime may not live long enough
  --> src/decode/libipt/mod.rs:51:9
   |
40 |     fn iter_blocks<'t>(
   |                    -- lifetime `'t` defined here
...
51 |         Box::new(itr)
   |         ^^^^^^^^^^^^^ returning this value requires that `'t` must outlive `'static`
   |
help: to declare that the trait object captures data from argument `self`, you can add an explicit `'t` lifetime bound
   |
43 |     ) -> Box<dyn Iterator<Item = Result<Block, HWTracerError>> + 't> {
   |                                                                ++++
```

Reply:
So maybe `'t` isn't needed but `'_` on the `Result` is?

Reply:
Like this?

```
-> Box<dyn Iterator<Item = Result<Block, HWTracerError>> + '_>
```

That's what we have currently.

Reply:
Without the `<'t>` and other `'t` stuff in the signature. [Maybe it's necessary but I think it might not be.]

Reply:
OK, sorry for the wild goose chase! I should learn not to be too clever with the borrow checker :)

Reply:
I tried the very same things myself when preparing this PR. The joys of Rust...